packages/react-reconciler/src/ReactFiberHydrationContext.js

```diff
@@ -855,14 +868,22 @@ function popHydrationState(fiber: Fiber): boolean {
       }
     }
   }
-  popToNextHostParent(fiber);
-  if (fiber.tag === SuspenseComponent) {
-    nextHydratableInstance = skipPastDehydratedSuspenseInstance(fiber);
-  } else {
-    nextHydratableInstance = hydrationParentFiber
-      ? getNextHydratableSibling(fiber.stateNode)
-      : null;
+  switch (fiber.tag) {
+    case SuspenseComponent:
+      nextHydratableInstance = skipPastDehydratedSuspenseInstance(fiber);
+      break;
+    case HostSingleton:
+      nextHydratableInstance = getNextHydratableInSingleton(
+        fiber.stateNode,
+        nextRootOrSingletonHydratable,
+      );
+      break;
+    default:
+      nextHydratableInstance = hydrationParentFiber
+        ? getNextHydratableSibling(fiber.stateNode)
+        : null;
   }
+  popToNextHostParent(fiber);
```

Comment:
inverting this function to go after the nextHydratableInstance seems like doesn't change any meaningful semantics. technically the default case will be considering the prior hydrationParentFiber rather than the next one however I don't think we ever pop up to a null parent b/c we should hit the root fiber and then end hydration